microprofile/openapi/src/main/java/io/helidon/microprofile/openapi/MPOpenAPIBuilder.java

```diff
@@ -30,15 +44,106 @@
-        return indexView.get();
+    public synchronized List<FilteredIndexView> perAppFilteredIndexViews() {
+        if (perAppFilteredIndexViews == null) {
+            perAppFilteredIndexViews = buildPerAppFilteredIndexViews();
+        }
+        return perAppFilteredIndexViews;
+    }
+
+    private List<FilteredIndexView> buildPerAppFilteredIndexViews() {
+        /*
+         * The JaxRsCdiExtension knows about all the apps in the system. For each app find out the classes related to that
+         * app -- the application class itself and any resource classes reported by its getClasses() or getSingletons()
+         * methods -- and create a FilteredIndexView that will be used to restrict scanning to only those classes for that app.
+         */
+        JaxRsCdiExtension ext = CDI.current()
+                .getBeanManager()
+                .getExtension(JaxRsCdiExtension.class);
+
+        /*
+         * There are two cases that return a default filtered index view. Don't create it yet, just declare a supplier for it.
+         */
+        Supplier<List<FilteredIndexView>> defaultResultSupplier = () -> List.of(new FilteredIndexView(indexView.get(),
+                openAPIConfig.get()));
+
+        /*
+         * Each set in the list holds the classes related to one app.
+         */
+        List<JaxRsApplication> appsToRun = ext.applicationsToRun();
+        if (appsToRun.size() <= 1) {
+            return defaultResultSupplier.get();
+        }
+        /*
+         * Some JaxRsApplication instances might have an application instance already associated with them. Others might not in
+         * which case we'll try to instantiate them ourselves (unless they are synthetic apps or lack no-args constructors).
+         */
+        List<Set<Class<?>>> appClassesToScan = appsToRun.stream()
+                .filter(MPOpenAPIBuilder::isNonSynthetic)
+                .map(this::appInstance)
+                .flatMap(Optional::stream)
+                .map(this::classesToScanForApp)
+                .collect(Collectors.toList());
+
+        if (appClassesToScan.size() <= 1) {
+            /*
+             * Use normal scanning with a FilteredIndexView containing no class restrictions (beyond what might already be in
+             * the configuration).
+             */
+            return defaultResultSupplier.get();
+        }
+        return appClassesToScan.stream()
+                .map(this::appRelatedClassesToFilteredIndexView)
+                .collect(Collectors.toList());
+    }
+
+    private static boolean isNonSynthetic(JaxRsApplication jaxRsApp) {
+        return !jaxRsApp.synthetic();
+    }
+
+    private Optional<? extends Application> appInstance(JaxRsApplication jaxRsApp) {
+        Application preexistingApp = jaxRsApp.resourceConfig().getApplication();
+        if (preexistingApp != null) {
+            return Optional.of(preexistingApp);
+        }
+        return instantiate(jaxRsApp.applicationClass());
```

Comment:
Please replace with 
```
return jaxRsApp.applicationClass()
                .flatMap(MPOpenAPIBuilder::instantiate);
```
And remove the `Optional<Class>` parameter from `instantiate` methods. Using Optional as a parameter is generally not recommended.